connector/connect/client/jvm/src/main/scala/org/apache/spark/ml/ConnectUtils.scala

```diff
@@ -0,0 +1,208 @@
+        builder.putParams(
+          param.name,
+          LiteralValueProtoConverter.toLiteralProto(value)
+        )
+      }
+    }
+    builder.build()
+  }
+
+  def serializeResponseValue(data: Any): proto.MlCommandResponse = {
+    data match {
+      case v: Vector => serializeVector(v)
+      case v: Matrix => serializeMatrix(v)
+      case _: Byte | _: Short | _: Int | _: Long | _: Float | _: Double | _: Boolean | _: String |
+           _: Array[_] =>
+        proto.MlCommandResponse
+          .newBuilder()
+          .setLiteral(LiteralValueProtoConverter.toLiteralProto(data))
+          .build()
+      case _ =>
+        throw new IllegalArgumentException()
+    }
+  }
+
+  def serializeVector(data: Vector): proto.MlCommandResponse = {
+    // TODO: Support sparse
+    val values = data.toArray
+    val denseBuilder = proto.Vector.Dense.newBuilder()
+    for (i <- 0 until values.length) {
+      denseBuilder.addValue(values(i))
+    }
+
+    proto.MlCommandResponse
+      .newBuilder()
+      .setVector(proto.Vector.newBuilder().setDense(denseBuilder))
+      .build()
+  }
+
+  def deserializeVector(protoValue: proto.Vector): Vector = {
+    // TODO: Support sparse
+    Vectors.dense(
+      protoValue.getDense.getValueList.stream().mapToDouble(_.doubleValue()).toArray
+    )
+  }
+
+  def deserializeMatrix(protoValue: proto.Matrix): Matrix = {
+    // TODO: Support sparse
+    val denseProto = protoValue.getDense
+    Matrices.dense(
+      denseProto.getNumRows,
+      denseProto.getNumCols,
+      denseProto.getValueList.stream().mapToDouble(_.doubleValue()).toArray
+    )
+  }
+
+  def serializeMatrix(data: Matrix): proto.MlCommandResponse = {
+    // TODO: Support sparse
+    // TODO: optimize transposed case
+    val denseBuilder = proto.Matrix.Dense.newBuilder()
+    val values = data.toArray
+    for (i <- 0 until values.length) {
+      denseBuilder.addValue(values(i))
+    }
+    denseBuilder.setNumCols(data.numCols)
+    denseBuilder.setNumRows(data.numRows)
+    denseBuilder.setIsTransposed(false)
+    proto.MlCommandResponse
+      .newBuilder()
+      .setMatrix(proto.Matrix.newBuilder().setDense(denseBuilder))
+      .build()
+  }
+
+  def deserializeResponseValue(protoValue: proto.MlCommandResponse): Any = {
+    protoValue.getMlCommandResponseTypeCase match {
+      case proto.MlCommandResponse.MlCommandResponseTypeCase.LITERAL =>
+        deserializeLiteral(protoValue.getLiteral)
+      case proto.MlCommandResponse.MlCommandResponseTypeCase.VECTOR =>
+        deserializeVector(protoValue.getVector)
+      case proto.MlCommandResponse.MlCommandResponseTypeCase.MATRIX =>
+        deserializeMatrix(protoValue.getMatrix)
+      case proto.MlCommandResponse.MlCommandResponseTypeCase.MODEL_REF =>
+        ModelRef.fromProto(protoValue.getModelRef)
+      case _ =>
+        throw new IllegalArgumentException()
+    }
+  }
+
+  def deserializeLiteral(protoValue: proto.Expression.Literal): Any = {
```

Comment:
@zhengruifeng Could you help move this utility function to "common" project ?

Reply:
a util function convert `proto.Expression.Literal` to `Any` value?
let me take a look

Reply:
https://github.com/apache/spark/pull/40485